diff --git a/package/boot/uboot-tools/uboot-envtools/files/qualcommax_ipq60xx b/package/boot/uboot-tools/uboot-envtools/files/qualcommax_ipq60xx
index a06ffc1f36..a6ff46ba33 100644
--- a/package/boot/uboot-tools/uboot-envtools/files/qualcommax_ipq60xx
+++ b/package/boot/uboot-tools/uboot-envtools/files/qualcommax_ipq60xx
@@ -13,6 +13,11 @@ case "$board" in
 cambiumnetworks,xe3-4)
 	ubootenv_add_mtd "0:APPSBLENV" "0x0" "0x10000" "0x10000"
 	;;
+jdcloud,re-cs-07)
+        mmcpart="$(find_mmc_part 0:APPSBLENV)"
+        [ -n "$mmcpart" ] && \
+            ubootenv_add_uci_config "$mmcpart" "0x0" "0x40000" "0x20000"
+	;;
 glinet,gl-ax1800|\
 glinet,gl-axt1800|\
 linksys,mr7350|\
diff --git a/target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-re-cs-07.dts b/target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-re-cs-07.dts
new file mode 100644
index 0000000000..8f68a46708
--- /dev/null
+++ b/target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-re-cs-07.dts
@@ -0,0 +1,282 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+/dts-v1/;
+
+#include "ipq6018.dtsi"
+#include "ipq6018-ess.dtsi"
+#include "ipq6018-nss.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/leds/common.h>
+
+/ {
+	model = "JDCloud RE-CS-07";
+	compatible = "jdcloud,re-cs-07", "qcom,ipq6018";
+
+	aliases {
+		serial0 = &blsp1_uart3;
+		serial1 = &blsp1_uart6;
+
+		led-boot = &led_status_blue;
+		led-failsafe = &led_status_red;
+		led-running = &led_status_green;
+		led-upgrade = &led_status_red;
+
+		ethernet0 = &dp1;
+		ethernet1 = &dp2;
+		ethernet2 = &dp3;
+		ethernet3 = &dp4;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&tlmm 56 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_status_red: red {
+			label = "red:status";
+			gpios = <&tlmm 57 GPIO_ACTIVE_HIGH>;
+		};
+
+		led_status_green: green {
+			label = "green:status";
+			gpios = <&tlmm 58 GPIO_ACTIVE_HIGH>;
+		};
+
+		led_status_blue: blue {
+			label = "blue:status";
+			gpios = <&tlmm 79 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&tlmm {
+	gpio-reserved-ranges = <20 1>;
+
+	btuart_pins: btuart-pins {
+		mux {
+			pins = "gpio48", "gpio49";
+			function = "blsp5_uart";
+			drive-strength = <8>;
+			bias-pull-up;
+		};
+	};
+
+	mdio_pins: mdio-pins {
+		mdc {
+			pins = "gpio64";
+			function = "mdc";
+			drive-strength = <8>;
+			bias-pull-up;
+		};
+
+		mdio {
+			pins = "gpio65";
+			function = "mdio";
+			drive-strength = <8>;
+			bias-pull-up;
+		};
+	};
+
+	usb_pwr_pins: usb-pwr-pins {
+		mux {
+			pins = "gpio22";
+			function = "gpio";
+			bias-pull-down;
+			output-high;
+		};
+	};
+};
+
+&blsp1_uart6 {
+	status = "okay";
+
+	pinctrl-0 = <&btuart_pins>;
+	pinctrl-names = "default";
+};
+
+&sdhc_1 {
+	status = "okay";
+
+	/delete-property/ mmc-hs400-1_8v;
+	mmc-ddr-1_8v;
+	mmc-hs200-1_8v;
+};
+
+&mdio {
+	status = "okay";
+
+	pinctrl-0 = <&mdio_pins>;
+	pinctrl-names = "default";
+	reset-gpios = <&tlmm 75 GPIO_ACTIVE_LOW>;
+
+	ethernet-phy-package@0 {
+		compatible = "qcom,qca8075-package";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		reg = <0>;
+
+		qca8075_24: ethernet-phy@24 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <24>;
+		};
+
+		qca8075_25: ethernet-phy@25 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <25>;
+		};
+
+		qca8075_26: ethernet-phy@26 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <26>;
+		};
+
+		qca8075_27: ethernet-phy@27 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <27>;
+		};
+	};
+};
+
+&switch {
+	status = "okay";
+
+	switch_lan_bmp = <(ESS_PORT2 | ESS_PORT3 | ESS_PORT4)>;
+	switch_wan_bmp = <ESS_PORT1>;
+	switch_mac_mode = <MAC_MODE_PSGMII>;
+
+	qcom,port_phyinfo {
+		port@1 {
+			port_id = <1>;
+			phy_address = <24>;
+		};
+
+		port@2 {
+			port_id = <2>;
+			phy_address = <25>;
+		};
+
+		port@3 {
+			port_id = <3>;
+			phy_address = <26>;
+		};
+
+		port@4 {
+			port_id = <4>;
+			phy_address = <27>;
+		};
+	};
+};
+
+&dp1 {
+	status = "okay";
+	phy-handle = <&qca8075_24>;
+	label = "wan";
+};
+
+&dp2 {
+	status = "okay";
+	phy-handle = <&qca8075_25>;
+	label = "lan1";
+};
+
+&dp3 {
+	status = "okay";
+	phy-handle = <&qca8075_26>;
+	label = "lan2";
+};
+
+&dp4 {
+	status = "okay";
+	phy-handle = <&qca8075_27>;
+	label = "lan3";
+};
+
+&usb3 {
+	pinctrl-0 = <&usb_pwr_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&edma {
+	status = "okay";
+};
+
+&CPU0 {
+	/delete-property/ cpu-supply;
+};
+
+&CPU1 {
+	/delete-property/ cpu-supply;
+};
+
+&CPU2 {
+	/delete-property/ cpu-supply;
+};
+
+&CPU3 {
+	/delete-property/ cpu-supply;
+};
+
+&rpm {
+	status = "disabled";
+};
+
+&blsp1_uart3 {
+	status = "okay";
+	pinctrl-0 = <&serial_3_pins>;
+	pinctrl-names = "default";
+};
+
+&crypto {
+	status = "okay";
+};
+
+&cryptobam {
+	status = "okay";
+};
+
+&pcie_phy {
+	status = "okay";
+};
+
+&prng {
+	status = "okay";
+};
+
+&qpic_bam {
+	status = "okay";
+};
+
+&qusb_phy_0 {
+	status = "okay";
+};
+
+&qusb_phy_1 {
+	status = "okay";
+};
+
+&ssphy_0 {
+	status = "okay";
+};
+
+&usb2 {
+	status = "okay";
+};
+
+&usb3 {
+	status = "okay";
+};
diff --git a/target/linux/qualcommax/image/ipq60xx.mk b/target/linux/qualcommax/image/ipq60xx.mk
index 7e38d384a8..e0ee6166fa 100644
--- a/target/linux/qualcommax/image/ipq60xx.mk
+++ b/target/linux/qualcommax/image/ipq60xx.mk
@@ -48,6 +48,21 @@ define Device/cambiumnetworks_xe3-4
 endef
 TARGET_DEVICES += cambiumnetworks_xe3-4
 
+define Device/jdcloud_re-cs-07
+	$(call Device/FitImage)
+	$(call Device/EmmcImage)
+	DEVICE_VENDOR := JDCloud
+	DEVICE_MODEL := ER1
+	DEVICE_DTS_CONFIG := config@cp03-c4
+	DEVICE_DTS := ipq6010-re-cs-07
+	SOC := ipq6018
+	DEVICE_PACKAGES := kmod-fs-ext4 mkf2fs f2fsck kmod-fs-f2fs
+	BLOCKSIZE := 64k
+	KERNEL_SIZE := 6144k
+	IMAGE/factory.bin := append-kernel | pad-to $${KERNEL_SIZE}  |  append-rootfs | append-metadata
+endef
+TARGET_DEVICES += jdcloud_re-cs-07
+
 define Device/glinet_gl-common
 	$(call Device/FitImage)
 	$(call Device/UbiFit)
diff --git a/target/linux/qualcommax/ipq60xx/base-files/etc/board.d/02_network b/target/linux/qualcommax/ipq60xx/base-files/etc/board.d/02_network
index b8b22bcffe..6f704f6976 100644
--- a/target/linux/qualcommax/ipq60xx/base-files/etc/board.d/02_network
+++ b/target/linux/qualcommax/ipq60xx/base-files/etc/board.d/02_network
@@ -21,6 +21,9 @@ ipq60xx_setup_interfaces()
 	cambiumnetworks,xe3-4)
 		ucidef_set_interface_lan "lan1 lan2" "dhcp"
 		;;
+	jdcloud,re-cs-07)
+		ucidef_set_interfaces_lan_wan "lan1 lan2 lan3" "wan"
+		;;
 	glinet,gl-ax1800|\
 	linksys,mr7350|\
 	linksys,mr7500|\
diff --git a/target/linux/qualcommax/ipq60xx/base-files/lib/upgrade/platform.sh b/target/linux/qualcommax/ipq60xx/base-files/lib/upgrade/platform.sh
index d31363521f..94b804a3c6 100644
--- a/target/linux/qualcommax/ipq60xx/base-files/lib/upgrade/platform.sh
+++ b/target/linux/qualcommax/ipq60xx/base-files/lib/upgrade/platform.sh
@@ -115,6 +115,11 @@ platform_do_upgrade() {
 		fw_setenv bootcount 0
 		nand_do_upgrade "$1"
 		;;
+	jdcloud,re-cs-07)
+	  CI_KERNPART="0:HLOS"
+	  CI_ROOTPART="rootfs"
+	  emmc_do_upgrade "$1"
+	  ;;
 	glinet,gl-ax1800|\
 	glinet,gl-axt1800|\
 	netgear,wax214|\
@@ -164,3 +169,11 @@ platform_do_upgrade() {
 		;;
 	esac
 }
+
+platform_copy_config() {
+  case "$(board_name)" in
+  jdcloud,re-cs-07)
+    emmc_copy_config
+    ;;
+  esac
+}
diff --git a/target/linux/qualcommax/ipq60xx/target.mk b/target/linux/qualcommax/ipq60xx/target.mk
index 06a85ab5e4..7f20467f80 100644
--- a/target/linux/qualcommax/ipq60xx/target.mk
+++ b/target/linux/qualcommax/ipq60xx/target.mk
@@ -1,6 +1,6 @@
 SUBTARGET:=ipq60xx
 BOARDNAME:=Qualcomm Atheros IPQ60xx
-DEFAULT_PACKAGES += ath11k-firmware-ipq6018
+DEFAULT_PACKAGES += ath11k-firmware-ipq6018 nss-firmware-ipq6018
 
 define Target/Description
 	Build firmware images for Qualcomm Atheros IPQ60xx based boards.
